File: python/problem1456.py
class Solution:
    def maxVowels(self, s: str, k: int) -> int:
        index = [-1] * 26
        base = ord('a')
        index[ord('a') - base] = 0
        index[ord('e') - base] = 1
        index[ord('i') - base] = 2
        index[ord('o') - base] = 3
        index[ord('u') - base] = 4
        curSum = 0
        for i in range(k):
            idx = index[ord(s[i]) - base]
            if idx != -1:
                curSum += 1
        result = curSum
        for i in range(k, len(s)):
            left = index[ord(s[i - k]) - base]
            right = index[ord(s[i]) - base]
            if left == -1 and right != -1:
                curSum += 1
                result = max(curSum, result)
            elif left != -1 and right == -1:
                curSum -= 1
        return result

File: Java/problem1456.java
class Solution {
    public int maxVowels(String s, int k) {
        int[] index = new int[26];
        Arrays.fill(index, -1);
        index['a' - 'a'] = 0;
        index['e' - 'a'] = 1;
        index['i' - 'a'] = 2;
        index['o' - 'a'] = 3;
        index['u' - 'a'] = 4;
        char[] chars = s.toCharArray();
        int left = 0, right = 0;
        int sum = 0;
        for (; right < k; ++right) {
            int idx = index[chars[right] - 'a'];
            if (idx!= -1) {
                ++sum;
            }
        }
        int result = sum;
        for (; right < chars.length; ++right) {
            int leftIdx = index[chars[left] - 'a'];
            int rightIdx = index[chars[right] - 'a'];
            
            if (leftIdx != -1 && rightIdx == -1) {
                --sum;
            }
            else if (rightIdx != -1 && leftIdx == -1) {
                ++sum;
                result = Math.max(result, sum);
            }
            ++left;
        }
        return result;
    }
}
